src/aggregation/agg_req.rs

```diff
@@ -0,0 +1,86 @@
+//! Contains the aggregation request tree. Used to build an
+//! [AggregationCollector](super::AggregationCollector).
+//!
+//! [Aggregations] is the top level entry point to create a request, which is a `HashMap<String,
+//! Aggregation>`.
+//!
+//! # Example
+//!
+//! ```verbatim
+//! let agg_req_1: Aggregations = vec![
+//!     (
+//!         "range".to_string(),
+//!         Aggregation::Bucket(BucketAggregation {
+//!             bucket_agg: BucketAggregationType::RangeAggregation(RangeAggregationReq {                               
+//!                 field_name: "score".to_string(),
+//!                 buckets: vec![(3f64..7f64), (7f64..20f64)],
+//!             }),
+//!             sub_aggregation: Default::default(),
+//!         }),
+//!     )]
+//!     .into_iter()
+//!     .collect();
+//! ```
+
+use std::collections::HashMap;
+
+pub use super::bucket::RangeAggregation;
+use super::metric::{AverageAggregation, StatsAggregation};
+
+/// The top-level aggregation request structure, which contains [Aggregation] and their user defined
+/// names.
+///
+/// The key is the user defined name of the aggregation.
+pub type Aggregations = HashMap<String, Aggregation>;
+
+/// Aggregation request of [BucketAggregation] or [MetricAggregation].
+///
+/// An aggregation is either a bucket or a metric.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Aggregation {
+    /// Bucket aggregation, see [BucketAggregation] for details.
+    Bucket(BucketAggregation),
+    /// Metric aggregation, see [MetricAggregation] for details.
+    Metric(MetricAggregation),
+}
+
+#[derive(Clone, Debug, PartialEq)]
+/// BucketAggregations create buckets of documents. Each bucket is associated with a rule which
+/// determines whether or not a document in the falls into it. In other words, the buckets
+/// effectively define document sets. Buckets are not necessarily disjunct, therefore a document can
+/// fall into multiple buckets. In addition to the buckets themselves, the bucket aggregations also
```

Comment:
Great two lines...

"///In other words, the buckets
/// effectively define document sets. Buckets are not necessarily disjunct,"